kythe/cxx/verifier/souffle_entrysteam_support.cc

```diff
@@ -0,0 +1,159 @@
+  explicit KytheEntryStreamReadStream(
+      const std::map<std::string, std::string>& rw_operation,
+      souffle::SymbolTable& symbol_table, souffle::RecordTable& record_table)
+      : souffle::ReadStream(rw_operation, symbol_table, record_table),
+        raw_input_(STDIN_FILENO) {
+    // Check that the relation has the correct type.
+    if (typeAttributes.size() == kEntryElements &&
+        typeAttributes[0] == "r:vname" && typeAttributes[1] == "s:symbol" &&
+        typeAttributes[2] == "r:vname" && typeAttributes[3] == "s:symbol") {
+      const auto& record_info = types["records"]["r:vname"];
+      size_t arity = record_info["arity"].long_value();
+      const auto& types = record_info["types"];
+      if (arity == kVNameElements) {
+        relation_ok_ = true;
+        for (size_t i = 0; i < arity; ++i) {
+          if (types[i].string_value() != "s:symbol") {
+            relation_ok_ = false;
+            break;
+          }
+        }
+      }
+    }
+    if (!relation_ok_) {
+      fprintf(stderr, R"(error: bad relation for kythe_entrystream. Use:
+
+.type vname = [
+  signature:symbol,
+  corpus:symbol,
+  path:symbol,
+  root:symbol,
+  language:symbol
+]
+
+.decl entry(source:vname, kind:symbol, target:vname, value:symbol)
+
+Note that this is an experimental feature and this declaration may change.
+)");
+    }
+  }
+
+ protected:
+  souffle::Own<souffle::RamDomain[]> readNextTuple() override {
+    if (!relation_ok_) {
+      return nullptr;
+    }
+    google::protobuf::uint32 byte_size;
+    google::protobuf::io::CodedInputStream coded_input(&raw_input_);
+    coded_input.SetTotalBytesLimit(INT_MAX);
+    if (!coded_input.ReadVarint32(&byte_size)) {
+      return nullptr;
+    }
+    coded_input.PushLimit(byte_size);
+    if (!entry_.ParseFromCodedStream(&coded_input)) {
+      fprintf(stderr, "error reading input stream\n");
+      return nullptr;
+    }
+    auto tuple = std::make_unique<souffle::RamDomain[]>(kEntryElements);
+    souffle::RamDomain vname[kVNameElements];
+    CopyVName(entry_.source(), vname);
+    tuple[kSourceEntry] = recordTable.pack(vname, kVNameElements);
+    tuple[kKindEntry] = symbolTable.unsafeLookup(entry_.fact_name());
+    CopyVName(entry_.target(), vname);
+    tuple[kTargetEntry] = recordTable.pack(vname, kVNameElements);
+    tuple[kValueEntry] = symbolTable.unsafeLookup(entry_.fact_value());
+    return tuple;
+  }
+
+ private:
+  /// Copies the fields from `vname` to `target`.
+  void CopyVName(const kythe::proto::VName& vname, souffle::RamDomain* target) {
+    target[kSignatureEntry] = symbolTable.unsafeLookup(vname.signature());
+    target[kCorpusEntry] = symbolTable.unsafeLookup(vname.corpus());
+    target[kPathEntry] = symbolTable.unsafeLookup(vname.path());
+    target[kRootEntry] = symbolTable.unsafeLookup(vname.root());
+    target[kLanguageEntry] = symbolTable.unsafeLookup(vname.language());
+  }
+  /// This entry is reused for each tuple.
+  kythe::proto::Entry entry_;
```

Comment:
It seem like this should just be a function-local in readNextTuple? is there any benefit to re-using the same Entry proto?